src/contracts/exchange/NiftyswapExchange20.sol

```diff
@@ -0,0 +1,810 @@
+
+    return currencyBought;
+  }
+
+  /**
+   * @dev Pricing function used for converting Tokens to currency token.
+   * @param _assetSoldAmount    Amount of Tokens being sold.
+   * @param _assetSoldReserve   Amount of Tokens in exchange reserves.
+   * @param _assetBoughtReserve Amount of currency tokens in exchange reserves.
+   * @return price Amount of currency tokens to receive from Niftyswap.
+   */
+  function getSellPrice(
+    uint256 _assetSoldAmount,
+    uint256 _assetSoldReserve,
+    uint256 _assetBoughtReserve)
+    override public pure returns (uint256 price)
+  {
+    //Reserves must not be empty
+    require(_assetSoldReserve > 0 && _assetBoughtReserve > 0, "NiftyswapExchange20#getSellPrice: EMPTY_RESERVE");
+
+    // Calculate amount to receive (with fee)
+    uint256 _assetSoldAmount_withFee = _assetSoldAmount.mul(FEE_MULTIPLIER);
+    uint256 numerator = _assetSoldAmount_withFee.mul(_assetBoughtReserve);
+    uint256 denominator = _assetSoldReserve.mul(1000).add(_assetSoldAmount_withFee);
+    return numerator / denominator; //Rounding errors will favor Niftyswap, so nothing to do
+  }
+
+  /***********************************|
+  |        Liquidity Functions        |
+  |__________________________________*/
+
+  /**
+   * @notice Deposit less than max currency tokens && exact Tokens (token ID) at current ratio to mint liquidity pool tokens.
+   * @dev min_liquidity does nothing when total liquidity pool token supply is 0.
+   * @dev Assumes that sender approved this contract on the currency
+   * @dev Sorting _tokenIds is mandatory for efficient way of preventing duplicated IDs (which would lead to errors)
+   * @param _provider      Address that provides liquidity to the reserve
+   * @param _tokenIds      Array of Token IDs where liquidity is added
+   * @param _tokenAmounts  Array of amount of Tokens deposited corresponding to each ID provided in _tokenIds
+   * @param _maxCurrency   Array of maximum number of tokens deposited for each ID provided in _tokenIds.
+   *                       Deposits max amount if total liquidity pool token supply is 0.
+   * @param _deadline      Timestamp after which this transaction will be reverted
+   */
+  function _addLiquidity(
+    address _provider,
+    uint256[] memory _tokenIds,
+    uint256[] memory _tokenAmounts,
+    uint256[] memory _maxCurrency,
+    uint256 _deadline)
+    internal nonReentrant()
+  {
+    // Requirements
+    require(_deadline >= block.timestamp, "NiftyswapExchange20#_addLiquidity: DEADLINE_EXCEEDED");
+
+    // Initialize variables
+    uint256 nTokens = _tokenIds.length; // Number of Token IDs to deposit
+    uint256 totalCurrency = 0;          // Total amount of currency tokens to transfer
+
+    // Initialize arrays
+    uint256[] memory liquiditiesToMint = new uint256[](nTokens);
+    uint256[] memory currencyAmounts = new uint256[](nTokens);
+    uint256[] memory tokenReserves = new uint256[](nTokens);
```

Comment:
I think `tokenReserves` doesn't have to be created before calling `_getTokenReserves`

Reply:
Will update